firebase-firestore/src/main/java/com/google/firebase/firestore/model/Document.java

```diff
@@ -36,59 +40,107 @@
   }
 
   private static final Comparator<Document> KEY_COMPARATOR =
-      new Comparator<Document>() {
-        @Override
-        public int compare(Document left, Document right) {
-          return left.getKey().compareTo(right.getKey());
-        }
-      };
+      (left, right) -> left.getKey().compareTo(right.getKey());
 
   /** A document comparator that returns document by key and key only. */
   public static Comparator<Document> keyComparator() {
     return KEY_COMPARATOR;
   }
 
-  private final ObjectValue data;
-
-  private final DocumentState documentState;
+  /** Creates a new Document from an existing ObjectValue. */
+  public static Document fromObjectValue(
+      DocumentKey key, SnapshotVersion version, ObjectValue data, DocumentState documentState) {
+    return new Document(key, version, documentState) {
```

Comment:
This new type doesn't really seem like it adds much. It's just the degenerate case where the proto has been pre-parsed. It seems like we could leave Document as a single class that just happens to have two constructors one that takes ObjectiValue as before and one that takes no ObjectValue but a proto and a serializer.

Reply:
My goal was to not inflate the current Document class by letting it know about the Serializer. I updated the class to now take a Function that deserializes Proto values.

FWIW, in my ideal world the ObjectValue-backed Document would go away - but that is many sprints from now.